<a href="https://colab.research.google.com/github/minhhuong05/Econometrics_Midterm_Assignment/blob/main/TestOutliner.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score, mean_squared_error
import numpy as np

# Load the CSV data into a pandas DataFrame
df = pd.read_csv('/content/bank-additional-full.csv', sep=';')
df = df.drop_duplicates()

# Convert the target variable from string to binary (0 or 1)
df['y'] = df['y'].map({'no': 0, 'yes': 1})

# Select the relevant features and target variable
features = ['campaign', 'emp.var.rate', 'cons.price.idx', 'euribor3m', 'nr.employed']
target = 'y'

X = df[features]
y = df[target]

# Split data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Create and train the model
model = LinearRegression()
model.fit(X_train, y_train)

# Make predictions on the test data
y_pred = model.predict(X_test)

# Evaluate the model's performance
r2 = r2_score(y_test, y_pred)
mse = mean_squared_error(y_test, y_pred)

print('R-squared:', r2)
print('Mean Squared Error:', mse)

# Test the model with a specific data point (optional)
new_data = {'campaign': 10, 'emp.var.rate': 2.5, 'cons.price.idx': -0.2, 'euribor3m': 1.2, 'nr.employed': -1000}  # Replace with your desired values
new_data_df = pd.DataFrame([new_data])

prediction = model.predict(new_data_df)[0]
print('Predicted value for the new data point:', prediction)

R-squared: 0.1282781451230075
Mean Squared Error: 0.0906567418966878
Predicted value for the new data point: 13.07670958951811


#Nhận xét
ta có thể đưa ra một số nhận xét về các điểm ngoại lệ:

R-squared: Giá trị R-squared khá thấp (0.128), điều này có thể ngụ ý rằng mô hình không giải thích được một phần lớn sự biến thiên của dữ liệu. Điều này có thể là do sự hiện diện của các điểm ngoại lệ trong tập dữ liệu, khiến cho mô hình không thể tốt hơn.
Mean Squared Error (MSE): Giá trị MSE cũng khá cao (0.091), cho thấy sự chênh lệch lớn giữa các giá trị dự đoán và giá trị thực tế. Các điểm ngoại lệ có thể đóng góp vào việc tăng giá trị MSE này.
Giá trị dự đoán cho điểm dữ liệu mới: Giá trị dự đoán (13.076) cho điểm dữ liệu mới có vẻ rất cao so với phạm vi dự kiến của biến mục tiêu (thường là 0 hoặc 1 trong bài toán phân loại nhị phân). Điều này có thể là dấu hiệu của một điểm ngoại lệ.

In [ ]:
#LOẠI BỎ CÁC ĐIỂM NGOẠI LỆ
# Identify outliers using Z-score
from scipy import stats

# Calculate Z-score for each predicted value
df['y_pred'] = model.predict(X)
df['Z-score'] = np.abs(stats.zscore(df['y_pred']))

# Define a threshold for Z-score
threshold = 3

# Remove outliers from the DataFrame
df = df[df['Z-score'] < threshold]

# Drop unnecessary columns
df = df.drop(columns=['y_pred', 'Z-score'])

# Re-split data into features and target
X = df[features]
y = df[target]

# Re-split data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Re-train the model
model.fit(X_train, y_train)

# Make predictions on the test data
y_pred = model.predict(X_test)

# Evaluate the model's performance
r2 = r2_score(y_test, y_pred)
mse = mean_squared_error(y_test, y_pred)

print('R-squared:', r2)
print('Mean Squared Error:', mse)

R-squared: 0.1157176471849174
Mean Squared Error: 0.08263095354630819


#Nhận xét:
giá trị R-squared là 0.116, điều này ngụ ý rằng mô hình chỉ giải thích được khoảng 11.6% sự biến thiên của biến mục tiêu. Điều này cho thấy mô hình không giải thích được một phần lớn sự biến thiên của dữ liệu.MSE càng thấp thì mô hình càng chính xác. Trong trường hợp này, giá trị MSE là 0.083, điều này ngụ ý rằng sai số bình phương trung bình giữa giá trị dự đoán và giá trị thực tế trên tập dữ liệu kiểm tra là khoảng 0.083.
Tóm lại, các giá trị R-squared và MSE này cho thấy mô hình hồi quy tuyến tính không đạt được hiệu suất tốt trên tập dữ liệu kiểm tra.

#Dùng Zcore test Outliner


In [ ]:
import pandas as pd
from sklearn.linear_model import LinearRegression
from scipy import stats

# Đọc dữ liệu từ file CSV
data = pd.read_csv('/content/bank-additional-full.csv', sep =';')

# Tạo DataFrame từ dữ liệu
df = pd.DataFrame(data)

# Tính toán giá trị P(y) dựa trên mô hình linear regression đã cho
df['P(y)'] = 0.57497 - 0.00246*df['campaign'] + 0.01828*df['emp.var.rate'] - 5e-05*df['cons.price.idx'] - 0.00169*df['euribor3m'] - 0.01857*df['nr.employed']

# Tính Z-score cho P(y)
df['Z-score'] = stats.zscore(df['P(y)'])

# In ra tất cả các giá trị Z-score
print("Tất cả các giá trị Z-score:")
print(df['Z-score'])

# Xác định các điểm ngoại lệ dựa trên Z-score
outliers = df[(df['Z-score'] > 3) | (df['Z-score'] < -3)]

# In ra các điểm ngoại lệ
print("Các điểm ngoại lệ:")
print(outliers)

Tất cả các giá trị Z-score:
0       -0.321835
1       -0.321835
2       -0.321835
3       -0.321835
4       -0.321835
           ...   
41183    2.852928
41184    2.852928
41185    2.851064
41186    2.852928
41187    2.849199
Name: Z-score, Length: 41188, dtype: float64
Các điểm ngoại lệ:
Empty DataFrame
Columns: [age, job, marital, education, default, housing, loan, contact, month, day_of_week, duration, campaign, pdays, previous, poutcome, emp.var.rate, cons.price.idx, cons.conf.idx, euribor3m, nr.employed, y, P(y), Z-score]
Index: []

[0 rows x 23 columns]


Tất cả các giá trị Z-score: Các giá trị Z-score có phân phối từ -0.321835 đến 2.852928. Điều này ngụ ý rằng hầu hết các giá trị Z-score đều nằm trong khoảng chấp nhận được (-3 đến 3), chỉ có một số ít giá trị Z-score lớn hơn 3 hoặc nhỏ hơn -3.
Các điểm ngoại lệ: Tuy có một số giá trị Z-score lớn hơn 3 hoặc nhỏ hơn -3, tuy nhiên, sau khi kiểm tra, không có điểm nào được xác định là ngoại lệ dựa trên ngưỡng Z-score = 3.
Dựa vào kết quả này, có thể kết luận rằng không có điểm ngoại lệ được xác định trong dữ liệu của bạn dựa trên tiêu chuẩn Z-score = 3. Điều này có thể ngụ ý rằng phân phối của P(y) không có các giá trị ngoại lệ hoặc là các điểm ngoại lệ không đủ lớn để được xác định bằng phương pháp Z-score này.







